Feature 97 change owner of project

4w31ss3 opened the pull request:
closes #97 
closes #567 
---
<sup>Albert Weißert <albert.weissert@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

de-jcup commented on sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/usecases/admin/user/UseCaseAdministratorAssignsOwnerToProject.java:
```diff
@@ -0,0 +1,26 @@
+/* @formatter:off */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+@UseCaseDefinition(
```
As mentioned before: The usecase and text parts should be more like "UseCaseAdministratorChangesProjectOwner" or something like that.

4w31ss3 replied:
done

de-jcup reviewed (changes requested):
Hello Albert, good job! I still have some change requests, please look at my comments.

4w31ss3 commented on sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectAdministrationRestController.java:
```diff
@@ -95,53 +99,62 @@ public void createProject(@RequestBody @Valid ProjectJsonInput input) {
+    /* @formatter:off */
+	@UseCaseAdministratorAssignsOwnerToProject(@Step(number=1,name="Rest call",description="Administrator does call rest API to assign new owner",needsRestDoc=true))
+    @RequestMapping(path = AdministrationAPIConstants.API_ASSIGN_OWNER_TO_PROJECT, method = RequestMethod.POST, produces= {MediaType.APPLICATION_JSON_VALUE})
+    @ResponseStatus(HttpStatus.CREATED)
```
I chose CREATED because of consistency to `assignUserToProject`. From REST-API user perspective every POST-request should create or change something on server side. The user has no deeper insight into whether some internal resource is created or changed with assignOwner as well as with assignUser.

In my opinion, both requests should have the same ResponseStatus. OK would be also suitable, because in both cases no user facing resource is created but already existing resources are altered.

Should I change both to HttpStatus.OK?

4w31ss3 replied:
Issue #567 created. Will be implemented in this PR.

de-jcup replied:
OK.

4w31ss3 commented on sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectAssignOwnerService.java:
```diff
@@ -0,0 +1,104 @@
+        Project project = projectRepository.findOrFailProject(projectId);
+        User newOwner = userRepository.findOrFailUser(userId);
+
+        if (project.owner == newOwner) {
```
My bad. In Swift this would check for property equality and would always work because name would be unique and therefore sufficient.

Changed.

4w31ss3 commented on sechub-administration/src/test/java/com/daimler/sechub/domain/administration/project/ProjectAssignUserServiceTest.java:
```diff
@@ -0,0 +1,95 @@
+        project1.users.add(previousUser);
+
+        when(projectRepository.findOrFailProject("project1")).thenReturn(project1);
+        when(previousUser.getName()).thenReturn("previous");
```
done

4w31ss3 commented on sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectAdministrationRestControllerRestDocTest.java:
```diff
@@ -163,6 +168,26 @@ public void restdoc_delete_project() throws Exception {
 	}
 
+	@Test
+    @UseCaseRestDoc(useCase=UseCaseAdministratorAssignsOwnerToProject.class)
```
done

4w31ss3 commented on sechub-doc/src/test/java/com/daimler/sechub/restdoc/ProjectAdministrationRestControllerRestDocTest.java:
```diff
@@ -163,6 +168,26 @@ public void restdoc_delete_project() throws Exception {
+                post(https(PORT_USED).buildAdminAssignsOwnerToProjectUrl(PROJECT_ID.pathElement(), USER_ID.pathElement()), "projectId1", "userId1").
+                contentType(MediaType.APPLICATION_JSON_VALUE)
+                ).
+        andExpect(status().isCreated()).
```
As in my answer above, from a REST-API user perspective assign user and assign owner are similar operations. So in my opinion they should have similar behavior.

The url and method are very close:
`POST -> project/{projectId}/membership/{userId}`

`POST -> project/{projectId}/owner/{userId}`

So I would prefer having them both behaving the same way. Should I change both to HttpStatus.OK?

4w31ss3 commented on sechub-notification/src/main/java/com/daimler/sechub/domain/notification/owner/InformThatProjectHasNewOwnerNotificationService.java:
```diff
@@ -0,0 +1,73 @@
+            return;
+        }
+
+        Set<String> ccMailsSet = projectMessage.getUserEmailAdresses();
```
done

4w31ss3 commented on sechub-administration/src/main/java/com/daimler/sechub/domain/administration/project/ProjectAssignOwnerService.java:
```diff
@@ -0,0 +1,104 @@
+        User previousOwner = project.owner;
+        project.owner = newOwner;
+
+        newOwner.getProjects().add(project);
```
added integration tests. Done

de-jcup commented:
Just a remark:

I have seen inside your integration test `ProjectAdministrationScenario3IntTest.java`:
```
        assertUser(USER_1).isOwnerOf(PROJECT_1);
        assertUser(USER_1).isOwnerOf(PROJECT_2);
        assertUser(USER_1).hasOwnerRole();
```

I would prefer:
```
  assertUser(USER_1).
             isOwnerOf(PROJECT_1).
             isOwnerOf(PROJECT_2).
             hasOwnerRole();
```
We have here a full fluent API and it's more readable (more like a "telling a story".

But let's keep it as is. For the future: Please use the fluent API approach like in all other integration tests.

de-jcup reviewed (changes requested):
Could you please 
rename 
`UseCaseIdentifier.UC_ADMIN_ASSIGNS_OWNER_TO_PROJECT`
to 
`UseCaseIdentifier.UC_ADMIN_CHANGES_PROJECT_OWNER`

? 
We have already changed usecase class to `UseCaseAdministratorChangesProjectOwner`
and the identifiers are used  by documentation generators .

Additional: Please read my comment about the fluent API usage inside integration tests - but this optional.

All other parts are well done, so I will merge after the mentioned parts are done.

4w31ss3 commented:
@de-jcup:
- Renamed `UseCaseIdentifier.UC_ADMIN_ASSIGNS_OWNER_TO_PROJECT` to `UseCaseIdentifier.UC_ADMIN_CHANGES_PROJECT_OWNER`.

- Refactored `ProjectAdministrationScenario3IntTest` to fluent API.
